src/ws/metrics.rs

```diff
@@ -64,24 +163,122 @@ impl Procedure {
+    metrics.connected_clients_total_collector.inc();
+}
+
+/// Decrements the count of connected clients.
+pub async fn decrement_connected_clients(metrics: Arc<Metrics>) {
+    metrics.connected_clients_total_collector.dec();
+}
+
+/// Records updates sent on subscription.
+/// This increments the counter of updates sent
+/// on subscription based on the event type.
+pub async fn record_friendship_event_updates_sent(metrics: Arc<Metrics>, event: FriendshipEvent) {
+    metrics
+        .updates_sent_on_subscription_total_collector
+        .with_label_values(&[event.as_str()])
+        .inc();
+}
+
+/// Records the size of the incoming payload of a procedure call.
+/// This adds the size of the procedure call incoming payload to the
+/// histogram for the specified procedure.
+pub async fn record_in_procedure_call_size<T: prost::Message>(
+    metrics: Arc<Metrics>,
+    procedure: Procedure,
+    msg: &T,
+) {
+    let size = calculate_message_size(msg);
+    metrics
+        .in_procedure_call_size_bytes_histogram_collector
+        .with_label_values(&[procedure.as_str()])
+        .observe(size as f64);
+}
+
+/// Records the size of the outgoing payload of a procedure call.
+/// This adds the size of the procedure call outgoing payload to the
+/// histogram for the specified procedure and response code.
+async fn record_out_procedure_call_size(
+    metrics: Arc<Metrics>,
+    code: Option<WsServiceError>,
+    procedure: Procedure,
+    size: usize,
+) {
+    let code = map_error_code(code);
+    metrics
+        .out_procedure_call_size_bytes_histogram_collector
+        .with_label_values(&[code, procedure.as_str()])
+        .observe(size as f64);
+}
+
+/// Records the duration of a procedure call.
+/// This adds the duration of the procedure call to the
+/// histogram for the specified procedure and response code.
+async fn record_request_procedure_call_duration(
+    metrics: Arc<Metrics>,
+    code: Option<WsServiceError>,
+    procedure: Procedure,
+    start_time: Instant,
+) {
+    let code = map_error_code(code);
+    let duration = Instant::now().duration_since(start_time).as_secs_f64();
+    metrics
+        .procedure_call_duration_seconds_histogram_collector
+        .with_label_values(&[code, procedure.as_str()])
+        .observe(duration);
+}
+
+/// Records a procedure call, its duration and its outgoing payload size.
+pub async fn record_procedure_call_and_duration_and_size(
     metrics: Arc<Metrics>,
     code: Option<WsServiceError>,
     procedure: Procedure,
+    start_time: Instant,
+    size: usize,
 ) {
-    let code = match code {
+    record_procedure_call(metrics.clone(), code.clone(), procedure.clone()).await;
+    record_request_procedure_call_duration(
+        metrics.clone(),
+        code.clone(),
+        procedure.clone(),
+        start_time,
+    )
+    .await;
+    record_out_procedure_call_size(metrics, code, procedure, size).await;
+}
```

Comment:
What do you think about moving all of these functions to `impl` block of `Metrics` struct? By doing that, you can avoid cloning the Arc each time you want to record something (although cloning an Arc is cheap)